### PR TITLE
Tag permissions

### LINKED ISSUE #233: tag editing restricted by permission [3]
# Acceptance Criteria

* buffalo tag my only be added/removed by LS staff
* philadelphia tag may only be added/removed by Molly
* all other tags can be edited by anyone
* in tag edit mode, uneditable tags will be marked as disabled (greyed out? lock?)

# Definitions

* a tag may be open or closed
* if closed, it may only be edited by whitelisted users

# Implementation Notes

* `tags.yml` must now include an `access` field for every tag

* API for tag permissions checking will be something like:

 ```ruby
some_user.permissions.can_tag?("philly")
```
* to accomplish whitelisting, we will store a list of tag ids that any given user may edit, according to (something like) the following schema in the (new) `user_permissions` table:

| user_id | resource_type | access_rules  |
|--|--|--|
| 1           | "List"                | NULL                          | # indicates user 1 is a 'lister'
| 2           | "Tagging"         | "{ tag_ids: [ 1, 2, 3] }" | # indicates user 2 can tag ids w/ tag_id 1, 2, & 3

(which indicates user 1 is a "lister" and user 2 can tag id's w/ `tag_id` == 1, 2, or 3)

### PR DESCRIPTION
**Pull request by aguestuser**
This resolves #233 and satisfies its Acceptance Criteria.
___

# Artifacts:

**Editing tags without permissions for nyc, philly, georgia:**

disabled-tags

**Granting editing permissions for nyc, philly, georgia:**

adding-permissions

**Editing tags with permissions:** 

editable-tags

# TODO

some random refactoring notes (writing them down so i don't forget them -- don't have to do all of these):

* would be nice to have a rake task to assign permissions to people who are suppposed to have them. or a bash script? (@aepyornis: thoughts?)
* generalize interface to `Permissions` class by introducing `permissions_for` (conceal and encapsulate `#tag_permissions`, `#list_permissions`, etc...
* (possibly) group resource-type-specific helpers (`#view_list`, `#edit_list`, `#owns_tag`, etc...) inside resource-specific `AccessRules` classes 
* resolve namespace issues by moving most of `Permissions` class into `UserPermissions` module,  renaming the module to `PermissionsBridge` (as per [this pattern](https://en.wikipedia.org/wiki/Bridge_pattern)) or perhaps `PermissionsAdapter` and (posisbly) renaming `UserPermission` to `Permission`
* put all of public `Permission` methods directly on `User` (only 3 left aside from legacy helpers: `#permissions_for`, `#add_permission` and `#remove_permission`)
* make `Tag` a proper model (possibly an `ActiveModel`?)

### CODE REVIEW

**Comment by aguestuser** [comment]
hmm... realizing now that i did *not* do the part of this card that involved removing the admin-only guard around tag views -- @aepyornis perhaps we can circle back to that next week? leave it admin-only for now while we gather feedback?

File: app/models/concerns/user_permissions.rb [hunk 1]
```diff
@@ -72,6 +98,17 @@ def list_permissions(list)
 
     private
 
+    # ACCESS RULE HELPER
+    def update_permission(resource_type, access_rules, operation)
+      permission = @user.user_permissions.find_or_create_by(resource_type: resource_type.to_s)
+      klass = "UserPermissions::#{resource_type}AccessRules".constantize
+      new_access_rules = klass.update(permission.access_rules, access_rules, operation).to_json
```
Comment (aepyornis) [inline app/models/concerns/user_permissions.rb]: I think some comments or refactoring here might help explain what's going on, and/or moving some logic to the active record model UserPermission.

This makes two similar #update calls: one is klass.update (which at this time is always TagAccessRules) and Permission.update (which executes sql via the @user.user_permissions model). Perhaps this entire helper function can be moved to the UserPermission model?

File: app/models/user_permission.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class UserPermission < ActiveRecord::Base
+  belongs_to :user
+  validates_presence_of :user_id, :resource_type
+
+  def access_rules
+    rules = super
+    rules && HashWithIndifferentAccess.new(JSON.parse(rules))
```
Comment (aepyornis) [inline app/models/user_permission.rb]: we can eliminate the JSON parse and to_json calls by telling rails to serialize the column as a hash:
``` rb
class UserPermission < ActiveRecord::Base
   serialize :access_rules, Hash
end
```

File: app/models/user_permission.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  end
+
+  def resource_type
+    super&.constantize
```
Comment (aepyornis) [inline app/models/user_permission.rb]: works for now...although we might have to change later to accommodate the legacy symfony permissions...

**Comment by aepyornis** [comment]
Great! 

This can be merged even if we don't address my comments.

I know we've gone back and forth about naming and if permissions should be a class or not, but after a few days, i have smallish suggestion that might help us:

class **UserPermission** : stays the same. 

module **UserPermissions** is removed entirely and all the methods can be placed directly on the User model. It's hard to imagine that the module could ever be used with another model besides user as currently written. it was only implemented as module to group together user functions about permssions, but those benefits are lost by the name confusion. 

class **UserPermisions::Permissions**  is moved to app/utility/permisisons.rb. Utility is where we keep various other helper classes and right now there's plenty of code in ``` Permisisons ``` to deserve it's own file and top-level namespace. 

If we do those changes we are left with only two similarly named thing: UserPermission (the active record model) in models/ and Permissions (a custom class) in utility.  

and thanks for all those refactoring notes! very useful.  Let's not forget about them in about a month or so when we get around to removing all the legacy permissions when the transition from symfony is complete.

File: app/models/concerns/user_permissions.rb (hunk 1 above)
Comment (aguestuser) [reply app/models/concerns/user_permissions.rb]: as discussed, i think this is clumsy and that there's a neat refactor to move `#update` onto the `UserPermission` model (which would already know its own `resource_type` and could statically enumerate its own `access_rule` logic sets), but we think we should put off this refactor (1) so we can keep moving, (2) b/c we are not touching tag permissions anytime soon, (3) b/c this refactor properly belongs within a larger refactor that moves more logic to `UserPermission` and (4) we want to handle permissions refactoring in one fell swoop as part of cleaning up after killing symphony.

File: app/models/user_permission.rb (hunk 2 above)
Comment (aguestuser) [reply app/models/user_permission.rb]: yes lets!

File: app/models/user_permission.rb (hunk 3 above)
Comment (aguestuser) [reply app/models/user_permission.rb]: +1 (but postponing for now...)

**Comment by aguestuser** [comment]
Looks good. Let's do this before merging: 

* do the cosmetic part of the `UserPermission` / `UserPermissions` / `Permissions` cleanup suggest above
* **but:** punt on moving any permissions logic between classes
* fix `all`/`byId` conflict
* add  `serialize :access_rules, Hash` helper to `UserPermission` as suggested above
* add link to refactoring discussion here in comment in `permissions.rb`

**Comment by aguestuser** [comment]
Let's also: 

* remove the `admin_only?` flag in views
* restrict to admins through a rake task

**Pull request merged**